### ISSUE TITLE
Force Pipe To Be Called

### ISSUE THREAD

**Issue by connorwyatt** [angular/angular#15041]
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Currently I have a stateful pipe that wraps a service and subscribes to an observable, and then stores the value and returns it every time the pipe's transform function is called. The issue is that I have to make the pipe impure and so it gets called on each change detection cycle.

**Expected behavior**
What I would like to be able to do is mark the pipe as dirty when I know the data has changed.

I have injected the change detector ref into the pipe, and have marked it for check and then detected changes, however none of the inputs have changed so the pipe doesn't get called again.

Because I know when the data has changed (as it is whenever the "next" function used to subscribe to the observable is called) I would like to be able to set the pipe as dirty to force it to be reevaluated when I call detect changes.

I have trawled the docs and source and can't find this behaviour currently, and if it is agreed that it would be worth implementing I would be happy to create a PR with some guidance.

**What is the motivation / use case for changing the behavior?**
To stop stateful pipes being polled especially when we know when the data has changed, in favour of a more push-based mechanism.

* **Angular version:** 2.X

**Comment by robwormald** [comment]
I think you're probably looking for the `WrappedValue` helper: https://github.com/angular/angular/blob/f2adb2900d93677517215a8ab6ccc10d95a5df48/modules/%40angular/core/src/change_detection/change_detection_util.ts#L27-L50

We use it for a similar purpose in the async pipe: https://github.com/angular/angular/blob/master/modules/@angular/common/src/pipes/async_pipe.ts#L107

**Comment by connorwyatt** [comment]
Hi Rob, if I've understood correctly, WrappedValue is used to create a new object reference so that the change detection detects a change to an input. What I need to do is to be able to trigger the pipe to execute again even though my inputs haven't changed.

**Comment by DzmitryShylovich** [comment]
@connorwyatt can you add a small plunkr that illustrates what you are truing to accomplish?

**Comment by connorwyatt** [comment]
Not really because it is an enhancement but I can try to give a scenario:

I have a (pure) pipe:

```
@Pipe({
  name: 'testPipe'
})
export class TestPipe implements PipeTransform {
  private _latestValue: string;

  constructor(private _ref: ChangeDetectorRef) {}

  transform(input: string): any {
    getObservableFromString(input).subscribe(data => {
      this._latestValue = data;
      this._ref.markForCheck();
      this._ref.detectChanges(); // This won't work because the input hasn't changed so the change detector will not reevaluate the pipe, but inside the pipe we know that the value has changed
    });
  }
}
```

In this example above (which is very similar to the Async pipe) I don't use `pure: false` as I don't want to check whether the value is the same every change detection cycle. What I want is some way of telling the change detector that I know something that it doesn't, namely that my data has changed. That way I can remove the `pure: false` and stop checking it all the time.

It's more of a proposal, and I'm not sure if it's possible or not as I know that the change detector ref in this case would belong to the parent component so you couldn't just mark it as dirty as it may have many children.

If I'm still not making myself clear let me know and I'll try to explain more.

Thanks

**Comment by DzmitryShylovich** [comment]
> this._ref.detectChanges(); // This won't work because the input hasn't changed so the change detector will not reevaluate the pipe, but inside the pipe we know that the value has changed

I don't think CDR has any effect on pipes at all :)

**Comment by connorwyatt** [comment]
That's my point, I want to be able to force the change detector to reevaluate it in this case.

**Comment by DzmitryShylovich** [comment]
have you tried to return `WrappedValue.wrap(this._latestValue);` instead of `this._ref.detectChanges();`?

**Comment by connorwyatt** [comment]
Again the issue there is that I would have to use `pure: false` otherwise my transform function won't be called.

What I'm trying to do ultimately is to stop Angular calling my transform function every change detection cycle. I want a way from inside of the pipe to basically say "Hey something internal changed! Run me again!".

**Comment by tytskyi** [comment]
@connorwyatt What you show above (https://github.com/angular/angular/issues/15041#issuecomment-285478197) is perfect use case for the attribute directive.

Why do you want to use a pipe for that?

**Comment by connorwyatt** [comment]
If I use an attribute directive then I can't pass the result into a component as a parameter, whereas if I use a pipe, I can do that. In this case I should be using a pipe as it transforms data into something else.

**Comment by meriturva** [comment]
I think this is one of the most important feature about pipes right now. I have a pipe that just check user roles compared to one passed on pipe arguments. Now i just would like to mark pipe for checking when user object change. I know i could resolve using impure pipe or using more args for pipe but i just would like to use role name as pipe input and user has service injected on pipe constructor. Force pipe to be checked when user changes is the only, elegant, solution i guess.

**Comment by MRamos62786** [comment]
https://plnkr.co/edit/Lw36FpsiGAoviJBdmZkD
Created a simple plunker to illustrate this. Clicking the new button, you'll note that the ngFor using the pipe doesn't work while just using a component member works fine. Using a component member is a fine workaround for this type of issue, though forcing the pipe would make for a cleaner solution.

**Comment by meriturva** [comment]
Hi to all,
i just would like to know if this important feature will be included on 5.x version. I think that find a way to trigger external update to pure pipe will have big positive impact on whole framework.

**Comment by mlc-mlapis** [comment]
@MRamos62786, @meriturva ... using `formCounter` on a pure pipe forces CD ... `<table *ngFor="let key of form.controls | keys:formCounter">`.

**Comment by stephenlautier** [comment]
I strongly agree with the trigger as it makes it much more performant and cleaner.

Currently, the best workaround I found is to make it as follows (maybe it helps anyone in the meantime)

```ts

@Pipe({
    name: "translate",
    pure: false
})
export class TranslationFallbackPipe implements PipeTransform, OnDestroy {

    private markForTransform = true;
    private value: string;

    private changes$$: Subscription;

    constructor(
        private service: TranslationService,
    ) {
       // NOTE: ngOnInit doesnt get triggered for pipes, thats why its in the ctor
        this.changes$$ = this.service.onTranslationChange$.subscribe(() => {
            this.markForTransform = true;
        });
    }

    transform(key: string, data?: object): string {
        if (!this.markForTransform) {
            return this.value;
        }
        this.value = this.service.getWithFallback(key, data);
        this.markForTransform = false;
        return this.value;
    }

    ngOnDestroy() {
        if (this.changes$$) {
            this.changes$$.unsubscribe();
        }
    }

}
```

**Comment by meriturva** [comment]
@stephenlautier thanks for your contribution, but main aspect here is about pure pipe. Pure pipe with an elegant way to trigger changes from a global service is a win for all.

**Comment by rolandoldengarm** [comment]
I once made the mistake to create an impure pipe to do a custom date transform. The performance went from smooth to absolutely terrible.
Never use impure pipes!
Fixed now by creating a directive :)

**Comment by samherrmann** [comment]
As a workaround I was able to get the desired behaviour using a directive as follows:
```
<span [myPseudoPipe]="prop" (change)="value=$event">{{ value }}</span>
```

This approach also allows for the resulting value to be piped into another pipe:
```
<span [myPseudoPipe]="prop" (change)="value=$event">{{ value | someOtherPipe }}</span>
```
The requested feature for pipes would significantly clean up this code though.

**Comment by samherrmann** [comment]
Ignore my last post. I use Angular CLI in my project, and the solution I posted works in a dev build, but when you do a production build you get an error that there is no property `value` in the component because `value` was never declared in the component class.

**Comment by meriturva** [comment]
Any news here? what about Angular 6.0 upcoming release?

**Comment by ng-ankit-bhardwaj** [comment]
https://medium.com/@ankitbhardwaj_49963/impure-pipes-in-angular-94ba7f112e4a

**Comment by connorwyatt** [comment]
Thanks for posting your article, but I don't think it adds anything to this discussion, we're not discussing what impure pipes are, but instead a way to improve performance by being able to tell the change detector that something has changed.

**Comment by djidel** [comment]
Any updates on this issue?
Implementing a solution for this will be a great performance benefit compared to the use of impure pipes.
Maybe by finding a way to link the transform() calls to only one special observable instead of calling transform() on every change detection as in impure pipes.

**Comment by arron21** [comment]
@djidel  here is an approach that worked for me

`Example Array = ['s1', 's2', 'a1', 'a2']`
Problem: 
I needed to refresh my pipe because new data was being added to the Example Array that was being filtered via a pipe that takes in a String ( ex: I am filtering with 's", so only 's1' and 's2' are currently visible.
Now I push 's3' to the array, but its not showing up on the frontend so I still only see 's1' and 's2'  because the pipe has not refreshed!

Solution:
After the new data has been added to the Example Array do the following

First: make sure you have the ChangeDetectorRef imported
`constructor( private cdr: ChangeDetectorRef){}`

Our pipe in the html
`<div *ngFor="let example of ExampleArray | nameMatch:parameterFilter">{{example}}</div>`

the code that makes the pure pipe refresh
```
    this.ExampleArray.push('s3');
    const string = this.parameterFilter;
    this.parameterFilter = '';
    this.cdr.detectChanges();
    this.parameterFilter = string;
    this.cdr.detectChanges();
```

Now with the new ChangeDetectorRef code added we should see s1, s2, s3 all on the frontend. Let me know if this worked for you.

**Comment by connorwyatt** [comment]
@arron21 Agreed, this works, but it misses the point. You need to inject the ChangeDetectorRef into every component that uses the pipe, and if you miss it, the pipe doesn't work properly. It's not very DRY.

It also isn't very good separation of concerns. Why should my component have to know that the pipe requires me to force detect changes when I change its inputs?

**Comment by meriturva** [comment]
Actually we are discussing about pipe that has injected service that it triggers a changeDetection internally. No external args, no external ChangeDetectorRef call. Just a **pure pipe** that trigger new render phase itself (we know that there is just one pipe instance when it is pure).

Thinking about two cases (just examples):

1. User roles changes: so a `permissionPipe` could disable/enable a button when `userService` emits new permissions
2. User language changes: so a `tranlastionPipe` could show correct label when `languageService` emits new language code value

Community solutions resolve it in two different ways:

1. Impure pipe: we know about performance and memory issues related to impure approach (see https://github.com/ngx-translate/core/blob/master/projects/ngx-translate/core/src/lib/translate.pipe.ts )
2. Pure pipe with trigger arg: we don't want to add more complexity to our templates
(see https://robisim74.github.io/angular-l10n/quick-start/#first-scenario-you-only-need-to-translate-texts )

As i know official i18n solution will, and will not have any way to change languages runtime (sad story here https://github.com/angular/angular/issues/16477 ).

**Comment by arron21** [comment]
@connorwyatt You are correct, the component should not have to trigger the refresh.
It seems like the pure pipe doesn't actually watch the dataset that it is filtering. 

I wonder maybe if there is a way of passing in the original dataset into the pipe and having the change detection logic within the pipe
`<div *ngFor="let example of ExampleArray | nameMatch:parameterFilter:ExampleArray">{{example}}</div>`

then within the nameMatch pipe would it always have the latest values of ExampleArray? hmm. this does seem like a feature that should just work with purepipes

**Comment by destructive-dragon** [comment]
We are trying to localize our router-links and there just doesn't seem to be a way of doing this :(
Impure pipes are unacceptable, we have hundreds if not thousands of links that all need to be localized. I tried refreshing the AppComponent, but the router caches all components that it uses and there's no way to disable it.

**Comment by mlc-mlapis** [comment]
@destructive-dragon  ... so use pure pipe, something like ...
```
import { Pipe, PipeTransform } from '@angular/core';

@Pipe({
	name: 'fn',
	pure: true
})
export class FnPipe implements PipeTransform {

	// A template syntax: {{ templateValue | fn:componentMethodRef:fnArgument }}
	// ... should be transformed to: null.componentMethodRef(templateValue, fnArgument)
	transform(
		templateValue: any,
		fnReference: Function,
		...fnArguments: Array<any>
	): any {

		fnArguments.unshift(templateValue);

		// The function reference will NOT BE INVOKED IN THE COMPONENT CONTEXT.
		// This means that the function reference won't have a "this" context
		// during execution, unless the function is pre-bound to the component.
		return(fnReference.apply(null, fnArguments));
	}
}

```

**Comment by destructive-dragon** [comment]
@mlc-mlapis impure pipe has a heavy performance impact for the hundreds of links we have on our page, and that's not really worth it considering that the average user is gonna use this functionality maybe once in total.

**Comment by mlc-mlapis** [comment]
@destructive-dragon ... average user? Then probably we don't understand what you wanted at all. I thought that you wanted to use a method for translation in a template and run it only once on the beginning or when you change the language (by its key like 'en', 'de', 'cs', ...), that's why I showed the **PURE PIPE** for that which is used generally and also for that case.

**Comment by destructive-dragon** [comment]
@mlc-mlapis No, that's not what I wanted. I have a router link like this:
`<a [routerLink]="['/movie' | localize]" translate>WEBAPP_MOVIE</a>`
When the translation in my app changes, I need this routerLink to update. I don't have direct access to the translation in my component, because quite frankly I don't want to add the service into all of my 200+ components.

This is how it works with an impure pipe:
```
@Pipe({
  name: 'localize',
  pure: false
})
export class LocalizeRouterPipe implements PipeTransform {
  curLang: string;

  constructor(private translate: TranslateService, ref: ApplicationRef) {
    this.translate.onLangChange.subscribe((lang: LangChangeEvent) => {
      this.curLang = lang.lang;
    })
  }

  transform(value: any, args?: any): any {
    if(this.curLang) {
      return "/" + this.curLang + value
    }
    return value;
  }

}
```

And now I want this or a reasonable alternative without the huge performance penalty. Right now it seems the only alternative is to create a custom component to use instead of `[routerLink]` since for whatever reason the angular team decided to not give components the same performance penalties as Pipes.

**Comment by mlc-mlapis** [comment]
@destructive-dragon ... OK, then.

**Comment by SirCremefresh** [comment]
Is something happening on this topic?

**Comment by arron21** [comment]
It seems the the conversation has basically come to this:

Impure pipes do not detect changes from the dataset they are pipe(ing)

This sounds like an Angular Core issue, and not something we have been able to solve ourselves.

**Comment by mlc-mlapis** [comment]
@arron21 ... the design of `impure pipes` is to run always, so why are you talking about the change detection?

**Comment by djidel** [comment]
@mlc-mlapis because using impure pipes in a relatively big app has a negative performance impact.

**Comment by arron21** [comment]
@mlc-mlapis @djidel my bad I meant to say **PURE** pipes... its been a long day 😄

**Comment by bryanrideshark** [comment]
@connorwyatt , what's the consensus from the core team on this issue? Any chance it will see the light of day? Would make Pipes much more powerful, though I recognize it might require a big refactor in something deep inside Angular's guts.

**Comment by bryanrideshark** [comment]
@connorwyatt , what if we changed the proposal to add another type of Pipe, an "OnPush" pipe?

These types of pipes would have to explicitly tell Angular that their value has changed, so Angular would treat them as "pure", and then check them again if they explicitly tell it to do so. They would have to implement an interface.

````
@Pipe({
  name: 'mostlyFast',
  pure: true
})
export class MostlyFastPipe implements PipeTransform, PipeChangesRequired, OnInit {
  readonly ngPipeChangesRequired$ = new Subject<void>();

  speed = 'mostly fast';

  constructor(someService: SomeService) {
  }

  ngOnInit() {
    this.someService.getSomeValue$.subscribe((value) => {
      this.speed = value;
      // Tells Angular that this pipe needs to be checked again.
      this.ngPipeChangesRequired$.next();
    });
  }

  transform(name: string) {
    return `${name} is moving ${this.speed}`;
  }

}
````

**Comment by connorwyatt** [comment]
@bryanrideshark Haha I have no idea (I'm not part of the Angular team). I considered this to be pretty important from a performance point of view but it seems like it's not a priority to the Angular team. Bear in mind that at this point this issue is 18 months old, I don't think you're going to see any movement on it now.

I've not been using Angular for at least a year anyway now so I haven't been chasing this, your best bet is to direct your comments to someone on the Angular team.

**Comment by connorwyatt** [comment]
Best of luck

**Comment by bryanrideshark** [comment]
@robwormald , what do you think about this? I know it'll throw a monkey wrench in Ivy, but would be good, no?

People are abandoning Angular because this one issue hasn't been addressed! (Did I get that right, @connorwyatt? 😜 )

**Comment by connorwyatt** [comment]
@bryanrideshark Haha you're putting words in my mouth there 😄

**Comment by EkiGui** [comment]
Any news on this topic ? I think the *translator pipe* provided by @stephenlautier is a good use case: 
- You use pipes to handle i18n
- Text has been rendered (*transformed*) with language A
- User selects language B
- **How can you trigger a new transform operation from inside your pipe ?**

This sounds like a job for ChangeDetectorRef but this service only applies for components, right ?

**Comment by destructive-dragon** [comment]
We fixed it by switching to Vue.js which by default only executes pipes if they get changed.

**Comment by EkiGui** [comment]
Thanks for your input.
For those who want to stick with Angular, using a directive (instead of a pipe) did the trick for us 😉

**Comment by jodinathan** [comment]
I switched to a component for translations.
It is not only easier to maintain, we could also add markdown and other stuff as options to translations.

Pipes should really be used with very simple stuff that has no need for changes if the input does not update.

Atenciosamente, 

Jonathan Rezende

**Comment by swapnilp0x00** [comment]
@jodinathan @EkiGui 
I made my pipe accept a parameter and always passed `myService.appLang$ | async ` as the parameter value. This did the trick for me.
`<p> {{ value | myLangPipe: (myService.appLang$ | async) }}</p>`
whenever I change the language using `myService.appLang$.next('en')`, my pipe executes transform block.

**Comment by eduboxgithub** [comment]
Any news regarding this matter?

I'm also using ngx translate, but dates and currency are set with with Angular DatePipe and CurrencyPipe. 

Now my app needs to support multilingual. 

The part that uses ngx translate works great but not the parts that use angular pipes.

Angular Pipes does not refresh on language/locale change.

How can I override the default angular pipes behavior?

**Comment by meriturva** [comment]
> Any news regarding this matter?
> 
> I'm also using ngx translate, but dates and currency are set with with Angular DatePipe and CurrencyPipe.
> 
> Now my app needs to support multilingual.
> 
> The part that uses ngx translate works great but not the parts that use angular pipes.
> 
> Angular Pipes does not refresh on language/locale change.
> 
> How can I override the default angular pipes behavior?

The only way I see its to subscribe on language change:

```
onLangChange.subscribe((event: LangChangeEvent) => {
  // do something
});
```
and [markForCheck](https://angular.io/api/core/ChangeDetectorRef#markForCheck) the view or use a [wrapped value](https://angular.io/api/core/WrappedValue) for date instance.

Angular Date and Currency pipes are pure!

**Comment by borriej** [comment]
Bumped into this one today. Is it safe to say that we should use class ```function``` for this kind of filtering? and only use ```pipe``` for filtering when content is only being filtered once / loaded once?

Cause i have an observable$ that changes, and the pipe needs re-do ```transform```. It's only working with ```pure: false```

**Comment by mlc-mlapis** [comment]
@borriej Pure pipe really helps as they are controlled by key value and when this one has changed, then `transform` runs again.

**Comment by TKul6** [comment]
Sorry but I don't think solution like "let's use directive where a pipe should be used" is the right way to go here.

@connorwyatt please notice this issue has been open for 2 years and it still on going, people (including me) still want this feature. Not sure why it is still inactive, but the need exists (and as I'm browsing this thread most of the people share the same idea about the solution).

Please reconsider having this feature as a requirement.

At least for me, instead just write the name of the pipe, it takes me about 3-5 minutes to get this (already written) pipe to work. It's very frustrating knowing you need to do dirty tricks due to lacks in the platform you are using.

**Comment by connorwyatt** [comment]
> Sorry but I don't think solution like "let's use directive where a pipe should be used" is the right way to go here.
> 
> @connorwyatt please notice this issue has been open for 2 years and it still on going, people (including me) still want this feature. Not sure why it is still inactive, but the need exists (and as I'm browsing this thread most of the people share the same idea about the solution).
> 
> Please reconsider having this feature as a requirement.
> 
> At least for me, instead just write the name of the pipe, it takes me about 3-5 minutes to get this (already written) pipe to work. It's very frustrating knowing you need to do dirty tricks due to lacks in the platform you are using.

Just want to make it clear, at no point have I said that a directive is a solution here, that's why I created this issue.

I don't know what you expect me to do though, I have absolutely zero influence here, and I don't use Angular anymore so I'm not particularly motivated to try.

You need to @ someone from the Angular team

**Comment by jodinathan** [comment]
I've ditched most pipes.
Directives and components are just so much better and organized.

**Comment by mlc-mlapis** [comment]
The original idea was:

> What I would like to be able to do is mark the pipe as dirty when I know the data has changed.

And I still think that the logic of a key-value change (including an object reference) on `pure pipes` is enough to control re-running them.

**Comment by connorwyatt** [comment]
> The original idea was:
> 
> > What I would like to be able to do is mark the pipe as dirty when I know the data has changed.
> 
> And I still think that the logic of a key-value change (including an object reference, that is the same mechanism as `OnPush` strategy is using) on `pure pipes` is enough to control re-running them.

I think you may have missed the original point of the feature request, it's not that the input has changed, it's that something else that you know about in your app has changed and you want to re-render

**Comment by mlc-mlapis** [comment]
@connorwyatt Sure, but I suppose that there is some relation between, because if not there is no necessity to re-render it. Then the key is the logic of such relations and how to decide that it happens.

It the application code itself is responsible for such a change, then you can use immutable principles and change a reference of an object (when changing some data in totally different parts), for example, and use it to invoke re-running the pure pipe. It supposes some abstraction of seeing, what's behind, but I think it's still a very good way.

**Comment by connorwyatt** [comment]
> @connorwyatt Sure, but I suppose that there is some relation between, because if not, there is no necessity to re-render it. Then the key is the logic of such relations and how to decide that it happens.
> 
> If the application code itself is responsible for such a change, then you can use immutable principles and change a reference of an object (when changing some data in totally different parts), for example, and use it to invoke re-running the pure pipe. It supposes some abstraction of seeing, what's behind, but I think it's still a very good way.

Understood that you disagree, I think you have slightly misunderstood the initial feature request however and there are many others that would find such a feature useful

**Comment by mlc-mlapis** [comment]
@connorwyatt It was nothing directly against, for sure. Having another possibility to somehow re-run pipes would be still useful, especially in a reactive way. I wanted only to point out that there is already one mechanism, how to do it.

**Comment by connorwyatt** [comment]
> @connorwyatt It was nothing directly against, for sure. Having another possibility to somehow re-run pipes would be still useful, especially in a reactive way. I wanted only to point out that there is already one mechanism, how to do it.

Yep, I am aware of that and you need to inject whatever the it is you are reacting to in the pipe everywhere that you use the pipe which in my opinion is a code smell. The feature would alleviate that

**Comment by jmcarter17** [comment]
> @EkiGui
> _For those who want to stick with Angular, using a directive (instead of a pipe) did the trick for us_
> Can you elaborate on this?
> At some places, I have multiple pipes chained after language translation pipe.
> @jodinathan
> I made my pipe accept a parameter and always passed `myService.appLang$ | async ` as the parameter value. This did the trick for me.
> `<p> {{ value | myLangPipe: (myService.appLang$ | async) }}</p>`
> whenever I change the language using `myService.appLang$.next('en')`, my pipe executes transform block.

Yes, this is exactly what I needed. Thanks so much for this idea.
Works like a charm, and much better than making the pipe impure.

**Comment by jmcarter17** [comment]
> Bumped into this one today. Is it safe to say that we should use a class `function` for template filtering when the observable$ gets changed? and only use `pipe` for filtering when content is only being filtered once / loaded once?
> 
> Cause i have an observable$ that changes, and the pipe needs re-do `transform`. It's only working with `pure: false`

@borriej if you have an observable that should trigger the transform, instead of making the pips impure, use the solution mentioned by @swapnilp0x00 and pass a parameter which is the (observable$ | async) to your pipe.
Works perfectly.

**Comment by connorwyatt** [comment]
It also means you have to inject the service into every single component that needs to render translatable strings

**Comment by jmcarter17** [comment]
Yes, that is true.
Unfortunate, but until the pipe can "subscribe" to the observable itself instead of passing it as a parameter, that at least allow to have pure pipes.
I was very annoyed at having impure pipes and this solves "that" problem.

**Comment by connorwyatt** [comment]
A pipe can subscribe to the observable, it's just that you can't force a re-render

**Comment by jmcarter17** [comment]
Yes, that is what I meant

**Comment by WeslleyDeSouza** [comment]
> It also means you have to inject the service into every single component that needs to render translatable strings

you could add a subscription to the app.component which triggers the re-render for all subcomponents.

`      <ng-container *ngIf="show">
          <ng-progress></ng-progress>
          <router-outlet></router-outlet>       
      </ng-container>`

` languages.onChange.subscribe(()=>{
            this.show = false;

           // ...(show loading modal or something)
         
             setTimeout(()=>      this.show = true ,100) // 50-300
  })`

**Comment by shivamd20** [comment]
I was facing the same issue and I solved it by making the pipe return an observable.

```
@Pipe({
  name: 'testPipe'
})
export class TestPipe implements PipeTransform {
  private _latestValue: string;

  constructor(private _ref: ChangeDetectorRef) {}

  transform(input: string): any {
    getObservableFromString(input).subscribe(data => {
      this._latestValue = data;
      this._ref.markForCheck();
      this._ref.detectChanges(); // This won't work because the input hasn't changed so the change detector will not reevaluate the pipe, but inside the pipe we know that the value has changed
    });
  }
}
```
So basically  this can be re-written as 

```
@Pipe({
  name: 'testPipe'
})
export class TestPipe implements PipeTransform {
  private _latestValue: string;

  constructor(private _ref: ChangeDetectorRef) {}

  transform(input: string): any {
  return  getObservableFromString(input);
  }
}
```

and in the template 

` { val | testPipe | async  }  `

**Comment by HansKrywaa** [comment]
A this is still not really possible (after years) i ended up by using a Observable + Async Logic as well.

**Comment by jagthebeetle** [comment]
A number of people have mentioned directives as a workaround, but AFAICT they negate the advantage of pipe purity. Specifically, directives don't have an OnPush concept, so they don't solve the problem of reducing the number of times their wrapped transformation has to be computed (#14627).

**Comment by meriturva** [comment]
Thanks for your solution @shivamd20 but the main point here is performance and memory consumption.

Using a pure pipe following an impure pipe (`async` is impure, see https://github.com/angular/angular/blob/master/packages/common/src/pipes/async_pipe.ts#L81)
is not a so smart solution for me.

What I feel after many small projects and big enterprise applications released is that pure pipes are singleton and they don't have the responsibility to force refresh itself. The pure pipe has to remain simple and fast. The pipe doesn't need to have any reference to parent view.

When we have to reevaluate all pipes present on a complex view we have to rethink triggers and change detection.

**Comment by rocifier** [comment]
In my opinion pipes should never be stateful. You should manipulate your data via state management first before you get it to the template and use pipes.

**Comment by Ghostbird** [comment]
I had a single, relatively simple scenario where I had to do something similar. When a certain object _x_ changed, I had to re-run the pipe that transformed object _y_.
In the end, I just passed _x_ and _y_ both to the pipe, even though the `transform` function only acts on _y_.

**Comment by stalniy** [comment]
> In my opinion pipes should never be stateful. You should manipulate your data via state management first before you get it to the template and use pipes.

It's not only about data. It's about reaction to side effects in the app. Imagine you have an app with permission management. When user is logged in/out, you want to update (hide/show some elements on the screen). To do this you may use pipe like `hasAccess`:

```html
<button *ngIf="'create' | hasAccess">create</button>
```

This pipe is pure because it translates results from `PermissionsService` but it should be able to reevaluate itself when permissions are changed (e.g., user logged out or admin changed user's role). Without this, you either forced to use `impure` pipe or reload the whole app

**Comment by mlc-mlapis** [comment]
@stalniy The key point is to not have a constant like `'create'` but a property that is possible to change and that change invokes the re-evaluation. The change could be even a reference change, as an immutable simple object when you change only its reference, because it's still a change from the pure pipe point of view.

Not `<button *ngIf="'create' | hasAccess">create</button>`,
but `<button *ngIf="roleRef | hasAccess">create</button>`.

**Comment by stalniy** [comment]
It was an example.  permissions sometimes are not defined roles but on user actions and subjects (ABAC vs RBAC). Then they are immutable.

The only way to handle this is to inject permissions in “all” UI components and pass them to every check `’read’ | hasAccess: permissions`. It’s possible but not developer friendly.

**Comment by rocifier** [comment]
> @mlc-mlapis
> It was an example. permissions sometimes are not defined roles but on user actions and subjects (ABAC vs RBAC). Then they are immutable.
> 
> The only way to handle this is to inject permissions in “all” UI components and pass them to every check `’read’ | hasAccess: permissions`. It’s possible but not developer friendly.
> 
> **Update**
> For more complex scenarios, it will be a bit more ridiculous:
> 
> `’read’ | hasAccess: ‘Article’: permissions`

Right I see. In my opinion pipes are not designed for this. In this scenario I just inject login meta info into components that need it. The only advantage I see of this is it decouples that information slightly more. But does it really need to be more decoupled than a typescript interface?

**Comment by survirtual** [comment]
This seems like a common sense capability to have with pipes.  There are countless scenarios where such a capability would be useful.

As an example, take a currency pipe.  With this capability, could be simply: {{ currencyAmount | currency }} where the pipe subscribes to a service which sets the currency type.  Currently, you'd have to do something like {{ currencyAmount | currency: currencyType }} which seems ludicrous to me.  It's more verbose, adds additional developer concerns, and demands the injection of a service which knows what the currency is increasing code complexity.

I could say the same for a user roles pipe, a localization pipe, and just about any other formatter that relies on a global configuration.

How has this issue been open for over 4 years?  It should be a trivial implementation that adds common sense, necessary capabilities.

**Comment by rocifier** [comment]
> This seems like a common sense capability to have with pipes. There are countless scenarios where such a capability would be useful.
> 
> As an example, take a currency pipe. With this capability, could be simply: {{ currencyAmount | currency }} where the pipe subscribes to a service which sets the currency type & conversion rates. Currently, you'd have to do something like {{ currencyAmount | currency: currencyType }} which seems ludicrous to me. It's more verbose, adds additional developer concerns, and demands the injection of a service which knows what the currency is increasing code complexity.
> 
> I could say the same for a user roles pipe, a localization pipe, and just about any other formatter that relies on a global configuration.
> 
> How has this issue been open for over 4 years? It should be a trivial implementation that adds common sense, necessary capabilities.

I don't think it would be trivial to implement because it breaks tree shaking. I could be wrong, but the OP is asking for a way to suddenly update part of a view if some global configuration changes outside of the view's change detection tree. If that was done it could potentially have far-reaching consequences. Even changing one character of text from a pipe could cause a whole page to re-render differently in html.

**Comment by stalniy** [comment]
@rocifier if you place a component in place of pipe, and component will re-render based on global state, will introduce any unexpected behavior? Correct, NO. Then why it can work the same way in a pipe that relates to this specific view?

**Comment by rocifier** [comment]
> @rocifier if you place a component in place of pipe, and component will re-render based on global state, will introduce any unexpected behavior? Correct, NO. Then why it cant work the same way in a pipe that relates to this specific view?

The main reason I can think of is that there would be a bit of additional overhead doing that inside of angular. And for the vast majority of cases pipes don't need the ability to have change detection. However I agree that when they do, there isn't really a solution.

**Comment by mlc-mlapis** [comment]
When using pure pipes and some specific cases where you need more control over when the pipe is re-run, it's possible to use an object as the primary value. Changing that reference will rerun the pipe. It's the same pattern as the OnPush mechanism.

**Comment by survirtual** [comment]
> > This seems like a common sense capability to have with pipes. There are countless scenarios where such a capability would be useful.
> > As an example, take a currency pipe. With this capability, could be simply: {{ currencyAmount | currency }} where the pipe subscribes to a service which sets the currency type & conversion rates. Currently, you'd have to do something like {{ currencyAmount | currency: currencyType }} which seems ludicrous to me. It's more verbose, adds additional developer concerns, and demands the injection of a service which knows what the currency is increasing code complexity.
> > I could say the same for a user roles pipe, a localization pipe, and just about any other formatter that relies on a global configuration.
> > How has this issue been open for over 4 years? It should be a trivial implementation that adds common sense, necessary capabilities.
> 
> I don't think it would be trivial to implement because it breaks tree shaking. I could be wrong, but the OP is asking for a way to suddenly update part of a view if some global configuration changes outside of the view's change detection tree. If that was done it could potentially have far-reaching consequences. Even changing one character of text from a pipe could cause a whole page to re-render differently in html. So for that reason a pipe would have to be made a part of the tree instead of just a template of a component.

Why can't any interpolated template content involving a pipe needing to modify data based on external services just have a variable auto-wrapped around it with the aggregated output added to the change detection tree for the parent it's a child of?

Give pipes an additional interface to be implemented indicating it modifies data or when ChangeDetectorRef is injected into a pipe, have the compiler perform a wrapping of the interpolated content being transformed and isolate change detection to a specific subset.

**Comment by survirtual** [comment]
Or give the pipe some kind of output parameter which gets automatically passed back into the pipe transform function without any additional template syntax.  The output parameter is added to the change detection tree, and when it gets modified, the transform is re-run.

I could go on but I feel like there are several plausible mechanisms with minimal / no impact outside of this specific use case.

**Comment by jodinathan** [comment]
I have changed my opinion regarding this subject and now I think that pipes should keep the current behavior.  
We have changed some pipes to Components and it all makes sense.

**Comment by survirtual** [comment]
> I have changed my opinion regarding this subject and now I think that pipes should keep the current behavior.
> We have changed some pipes to Components and it all makes sense.

Okay, so any time I want to **format data** based on an external configured element, here's what I'll do.

1) I'll make a component, sort of like this one:
```typescript
import { Component, OnDestroy, Input } from '@angular/core';
import { Subscription } from 'rxjs';

@Component({
  selector: 'this-should-be-a-pipe',
  template: '{{ modifiedInput }}',
  styleUrls: ['./thisShouldBeAPipe.component.css']
})
export class ThisShouldBeAPipe implements OnDestroy {
    _data: string;

    @Input() set data(value: string) {
        this._data = value;
        this.refreshText();
    }
    
    formattedData: string;

    _subscription: Subscription;
  
    constructor( private configuringService: ConfiguringService) { 
        this._subscription = this.configuringService.onChange.subscribe((config) => {
            this.refreshText();
        })
    }

    refreshText() {
        this.formattedData = this.configuringService.formatMyData(this._data);
    }

    ngOnDestroy(): void {
        if (this._subscription != null) {
            this._subscription.unsubscribe();
        }
    }
}
```

2) Then in the template, all I have to do is:
```html
...
<this-should-be-a-pipe [data]="myData"></this-should-be-a-pipe>
...
```
Instead of:
```html
<span>{{ myData | pipe }} </span>
```

I mean, okay.  It just smells a bit, as I'm ultimately formatting text / otherwise transforming text and would like to keep that concern isolated to pipes.  The idea that I need to use components to perform a simple transformation based on configuration data is a bit odd.  I think that this circumstance is significantly more common than you give it credit for, especially when making an app for different audiences interacting with the same base data values, or making anything configurable ie, weights, speeds, date formats, etc....

Which is what I thought pipes dealt with.

**Comment by albronils** [comment]
So basically Angular tells me to use pipes in order to be performant and reduce calculations, but whenever I want calculations based on some not frequently changed, global setting, I have to either write extremely redundant code or sacrifice performance or sacrifice a valuable set of features of pipes. Not something to get over easily tbh..
I think a decent enough approach could be something like adding change detection of pipe getters.

@survirtual thank you for sharing this code, it really helped me finding a clean enough approach for my translations which I want to share (in an abstract manner using your wording if you don't mind).

```typescript
import { Component, Input } from '@angular/core';
import { Subscription } from 'rxjs';

@Component({
  selector: 'this-should-be-a-pipe',
  template: '{{ data | pipe:setting }}',
  styleUrls: ['./thisShouldBeAPipe.component.css']
})
export class ThisShouldBeAPipe {
    @Input() data: string = ""

    get setting(): SettingType {
         return this.configuringService.setting
    }
  
    constructor(private configuringService: ConfiguringService) {}
}
```

with the same usage 

```html
...
<this-should-be-a-pipe [data]="myData"></this-should-be-a-pipe>
...
```

it basically wraps the pipe, adding any additional params which are subject to change without violating clean code.
Hope it helps anybody. Unfortunately still misses out some essential features of pipes.. (like usage in attributes/directives)

**Comment by csisy** [comment]
This feature would be really useful, and it's still in "under consideration" since Jun 5 2021.

**Comment by stalniy** [comment]
I found a replacement for pipes, in situations like that we can replace pipes with services observables. For example, for permission checks:


```ts
@Component({
  selector: 'my-home',
  template: `
    <ng-container *ngIf="can$ | async as can">
      <h1>Home Page</h1>
      <button *ngIf="can('create', 'Post')">Create Post</button>
    </ng-container>
  `
})
export class HomeComponent {
  readonly can$: Observable<AppAbility>;

  constructor(abilityService: AbilityService<AppAbility>) {
    this.can$ = abilityService.can$;
  }
}
```

or for compound state:

```ts
@Component({
  selector: 'my-home',
  template: `
    <ng-container *ngIf="state$ | async as state">
      <h1>Home Page</h1>
      <button *ngIf="state.can('create', 'Product')">Create Product</button>
      <my-product *ngFor="let product of state.products" [product]="product" [canDelete]="state.can('delete', product)"></my-product>
    </ng-container>
  `
})
export class HomeComponent {
  readonly state$: Observable<{ can: AppAbility["can"]; products: Product[] }>;

  constructor(abilityService: AbilityService<AppAbility>, productsService: ProductsService) {
    this.state$ = combineLatest([
      abilityService.can$,
      productsService.getProducts()
    ]).pipe(map(([can, products]) => ({
      can,
      products
    }));
  }
}
```

The same works well with translation service. Basically we can create our own observables and use `async` pipe to extract them. This observables may emit values any time they need and as a result our template will be rechecked. Additionally, they can implement memoization, to speed things up.

This is kind of home-made pipes implementation that does not suffer from Angular's limitations for pipes.

**Comment by pkozlowski-opensource** [comment]
Looking into this (and similar) issues it sounds like we are looking for using a pipe as something akin to computed properties. We do consider a more generic solution. Since we are getting multiple issues / feature requests around computed properties, we've opened a canonical issue #47553 to track this use-case. Currently there is no exact solution / design, but we do acknowledge the use-cases and the need for such concept.
